File: core/src/ru/myitschool/dcrawler/skills/Bomb.java
package ru.myitschool.dcrawler.skills;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.math.DiceAction;
import ru.myitschool.dcrawler.math.MathAction;
import ru.myitschool.dcrawler.skills.action.Action;
import ru.myitschool.dcrawler.skills.action.SimpleAttackAction;
import ru.myitschool.dcrawler.skills.patterns.FloorSplashTargetPattern;
import ru.myitschool.dcrawler.skills.patterns.TargetPattern;
import ru.myitschool.dcrawler.skills.targeting.SplashDisplayer;
import ru.myitschool.dcrawler.skills.targeting.TargetDisplayer;

/**
 * Created by Voyager on 25.11.2017.
 */
public class Bomb extends Skill {

    MathAction attackAction = new DiceAction(4);

    public Bomb(Entity doer) {
        super(doer);
        addPlayContainer().getEntityPlay().addAction(new SimpleAttackAction(this, attackAction));
    }

    @Override
    public void maintainDisplayers(Array<TargetDisplayer> displayers) {
        displayers.add(new SplashDisplayer(true));
    }

    @Override
    public TargetPattern getPattern() {
        return new FloorSplashTargetPattern(this);
    }

    @Override
    public Texture getIcon() {
        return new Texture("bomb.png");
    }

    @Override
    public String getName() {
        return "Bomb";
    }

    @Override
    public String getDescription() {
        return "Deal " + attackAction.getDescription() + " damage to all creatures in small zone";
    }

    @Override
    public int getSkillAccuracyBonus() {
        return 0;
    }

    @Override
    public int getRange() {
        return 2;
    }

    @Override
    public int getDistanceMin() {
        return 2;
    }

    @Override
    public int getDistanceMax() {
        return 8;
    }

    @Override
    public int getTargetCountMax() {
        return 1;
    }

    @Override
    public int getCooldownMax() {
        return 0;
    }

    @Override
    public int getType() {
        return SKILL_TYPE_AT_WILL;
    }

    @Override
    public int getTargetType() {
        return SKILL_TARGET_TYPE_FLOOR_SPLASH;
    }

    @Override
    public boolean isCheckAllTargets() {
        return false;
    }

    @Override
    public boolean isMarkEverything() {
        return false;
    }

    @Override
    public boolean isMark() {
        return true;
    }

    @Override
    public boolean isObstruct() {
        return true;
    }

    @Override
    public boolean isWallTargets() {
        return true;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/skills/Bomb.kt
package com.wolg_vlad.dcrawler.entities.skills

import com.badlogic.gdx.graphics.Texture
import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.entities.skills.action.SimpleAttackAction
import com.wolg_vlad.dcrawler.entities.skills.patterns.FloorSplashTargetPattern
import com.wolg_vlad.dcrawler.entities.skills.patterns.TargetPattern
import com.wolg_vlad.dcrawler.entities.skills.targeting.SplashDisplayer
import com.wolg_vlad.dcrawler.entities.skills.targeting.TargetDisplayer
import com.wolg_vlad.dcrawler.math.DiceAction
import com.wolg_vlad.dcrawler.math.MathAction

/**
 * Created by Voyager on 25.11.2017.
 */
class Bomb(doer: Entity) : Skill(doer) {
    var attackAction: MathAction = DiceAction(4)

    override val displayers: Array<TargetDisplayer> = arrayOf(SplashDisplayer(true))
    override val pattern: TargetPattern = FloorSplashTargetPattern(this)
    override val icon: Texture = Texture("bomb.png")
    override val name: String = "Bomb"
    override val description: String = "Deal ${attackAction.description} damage to all creatures in small zone"
    override val skillAccuracyBonus: Int = 0
    override val range: Int = 2
    override val distanceMin: Int = 2
    override val distanceMax: Int = 8
    override val targetCountMax: Int = 1
    override val cooldownMax: Int = 0
    override val type: Int = SKILL_TYPE_AT_WILL
    override val targetType: Int = SKILL_TARGET_TYPE_FLOOR_SPLASH
    override val isCheckAllTargets: Boolean = false
    override val isMarkEverything: Boolean = false
    override val isMark: Boolean = true
    override val isObstruct: Boolean = true
    override val isWallTargets: Boolean = true

    init {
        addPlayContainer().entityPlay.addAction(SimpleAttackAction(this, attackAction))
    }
}